clients/src/main/java/org/apache/kafka/clients/admin/RemoveMemberFromGroupResult.java

```diff
@@ -19,67 +19,53 @@
 import org.apache.kafka.common.KafkaFuture;
 import org.apache.kafka.common.internals.KafkaFutureImpl;
 import org.apache.kafka.common.message.LeaveGroupRequestData.MemberIdentity;
-import org.apache.kafka.common.message.LeaveGroupResponseData.MemberResponse;
 import org.apache.kafka.common.protocol.Errors;
-import org.apache.kafka.common.requests.LeaveGroupResponse;
 
-import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 
 /**
- * Result of a batch member removal operation.
+ * The result of the {@link Admin#removeMemberFromConsumerGroup(String, RemoveMemberFromConsumerGroupOptions)} call.
+ *
+ * The API of this class is evolving, see {@link Admin} for details.
  */
 public class RemoveMemberFromGroupResult {
 
-    private final Errors topLevelError;
-    private final Map<MemberIdentity, KafkaFuture<Void>> memberFutures;
-    private boolean hasError = false;
-
-    RemoveMemberFromGroupResult(LeaveGroupResponse response,
-                                List<MemberIdentity> membersToRemove) {
-        this.topLevelError = response.topLevelError();
-        this.memberFutures = new HashMap<>(membersToRemove.size());
-
-        if (this.topLevelError != Errors.NONE) {
-            // If the populated error is a top-level error, fail every member's future.
-            for (MemberIdentity memberIdentity : membersToRemove) {
-                KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
-                future.completeExceptionally(topLevelError.exception());
-                memberFutures.put(memberIdentity, future);
-            }
-            hasError = true;
-        } else {
-            for (MemberResponse memberResponse : response.memberResponses()) {
-                KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
-                Errors memberError = Errors.forCode(memberResponse.errorCode());
-                if (memberError != Errors.NONE) {
-                    future.completeExceptionally(memberError.exception());
-                    hasError = true;
-                } else {
-                    future.complete(null);
-                }
-                memberFutures.put(new MemberIdentity()
-                                      .setMemberId(memberResponse.memberId())
-                                      .setGroupInstanceId(memberResponse.groupInstanceId()), future);
-            }
-        }
-    }
+    KafkaFuture<Map<MemberIdentity, Errors>> future;
 
-    public Errors topLevelError() {
-        return topLevelError;
+    RemoveMemberFromGroupResult(KafkaFuture<Map<MemberIdentity, Errors>> future) {
+        this.future = future;
     }
 
-    public boolean hasError() {
-        return hasError;
+    /**
+     * Returns a future which contains the top level error (if any).
```

Comment:
Did we decide that this API should fail if there are any member failures? I don't think the current logic implements that.

Reply:
We haven't decided yet. IMO, it's better to make sure we only complete normally when the request is 100% success

Reply:
I agree. In that case, I guess we'll have to iterate over the errors in the map here?

Reply:
We could return a separate future for `all()` call and use `whenComplete` to finish for the internal future to return.